File: practica 1/sesion 3/DS_S3_C++Filters/Trabajo.cpp
#include <exception>
#include <iostream>
using namespace std;

#include "Trabajo.h"
#include "Cliente.h"
#include "Tecnico.h"
#include "ParteTecnico.h"
#include "ParteTrabajo.h"

void Trabajo::actualizarPresupuesto(float unPrecio) {
	presupuesto = unPrecio ;
}

void Trabajo::darParteTecnico(int amabilidad, int profesionalidad, int eficacia) {
	if (parteTecnico == NULL){
		parteTecnico = new ParteTecnico(amabilidad,profesionalidad,eficacia);
	}
	else{
		parteTecnico->setAmabilidad(amabilidad);
		parteTecnico->setProfesionalidad(profesionalidad);
		parteTecnico->setEficacia(eficacia);
	}
}

void Trabajo::darParteTrabajo(int horasExtra, int materialesExtra, int desplazamiento) {
	if (parteTrabajo == NULL){
		parteTrabajo = new ParteTrabajo(horasExtra, materialesExtra, desplazamiento);
	}
	else{
		parteTrabajo->setCosteDesplazamiento(desplazamiento);
		parteTrabajo->setCostoMaterialesExtra(materialesExtra);
		parteTrabajo->setHorasExtra(horasExtra);
	}
}

float Trabajo::getPresupuesto(){
	return presupuesto;
}

Trabajo::Trabajo(const string & descripcion, Cliente * cliente_creador){
	this->descripcion = descripcion;
	this->cliente = cliente_creador;
	hecho = false;
}

string Trabajo::getDescripcion() {
	return descripcion;
}

Tecnico * Trabajo::getTecnico() const {
	return tecnico;
}

void Trabajo::asignarTecnico(Tecnico * nuevo_tecnico) {
	this->tecnico = nuevo_tecnico;
	nuevo_tecnico->asignarTrabajo(this);
}

void Trabajo::realizado(){
	if (tecnico) 	hecho = true;
	else cout << "El trabajo aun no tiene asignado un técnico" << endl;
}

ParteTecnico * Trabajo::getParteTecnico() const {
	return this->parteTecnico;
}

ParteTrabajo * Trabajo::getParteTrabajo() const {
	return this->parteTrabajo;
}

float Trabajo::setPresupuesto(const float & nuevoPresupuesto) {
	return this->presupuesto = nuevoPresupuesto;
}

File: practica 3/implementacion dart p3/lib/ModeloSOSEMADO/Trabajo.dart
import 'Cliente.dart';
import 'Tecnico.dart';
import 'ParteTecnico.dart';
import 'ParteTrabajo.dart';

class Trabajo {
  Cliente cliente;
  Tecnico tecnico;
  ParteTrabajo parteTrabajo;
  ParteTecnico parteTecnico;

  String descripcion;
  bool hecho;
  double presupuesto;

  Trabajo(String desc, Cliente clienteCreador) {
    this.descripcion = desc;
    this.cliente = clienteCreador;
  }

  void asignarTecnico(Tecnico nuevoTecnico) {
    tecnico = nuevoTecnico;
    nuevoTecnico.asignarTrabajo(this);
  }

  void realizado() {
    if (tecnico != null) {
      hecho = true;
    } else {
      print("El trabajo no tiene asignado un técnico\n");
    }
  }

  void setPresupuesto(double nuevoPresupuesto) {
    presupuesto = nuevoPresupuesto;
  }

  void darParteTecnico(int amable, int profesional, int eficaz) {
    if (parteTecnico != null) {
      parteTecnico.setAmabilidad(amable);
      parteTecnico.setProfesionalidad(profesional);
      parteTecnico.setEficacia(eficaz);
    } else {
      parteTecnico = new ParteTecnico(amable, profesional, eficaz);
    }
  }

  void darParteTrabajo(
      int horas, int costoMateriales, int costoDesplazamiento) {
    if (parteTrabajo != null) {
      parteTrabajo.setHorasExtra(horas);
      parteTrabajo.setCosteMaterialesExtra(costoMateriales);
      parteTrabajo.setCosteDesplazamiento(costoDesplazamiento);
    } else {
      parteTrabajo =
          new ParteTrabajo(horas, costoMateriales, costoDesplazamiento);
    }
  }

  double getPresupuesto() {
    return presupuesto;
  }

  String getDescripcion() {
    return descripcion;
  }

  Tecnico getTecnico() {
    return tecnico;
  }

  ParteTecnico getParteTecnico() {
    return parteTecnico;
  }

  ParteTrabajo getParteTrabajo() {
    return parteTrabajo;
  }
}
